Add support for SDL-compatible joypad GUID and `Input.get_joy_info()` on Android

Nintorch opened the pull request:
Related to https://github.com/godotengine/godot/issues/56181 (if this PR doesn't fix this issue, then we need to create a mapping for Xbox controllers ourselves)

This PR adds support for SDL-compatible joypad GUIDs on Android, so now we should be able to use the mappings from [SDL controller mapping database for Android controllers](https://github.com/godotengine/godot/blob/63227bbc8ae5300319f14f8253c8158b846f355b/core/input/gamecontrollerdb.txt#L1828) that start with `0500`, while still keeping the compatibility with mappings that use the older GUID format (if a mapping with the new format can't be found, the code tries to find one with the old format).
This PR adds support for `Input.get_joy_info()` for Android as well.

See also [SDL's `Android_AddJoystick`](https://github.com/libsdl-org/SDL/blob/ec1d66d4f693b367ff60f01eac74406b3615fbed/src/joystick/android/SDL_sysjoystick.c#L308C112-L308C123), [SDL's `SDLControllerManager`](https://github.com/libsdl-org/SDL/blob/release-3.2.x/android-project/app/src/main/java/org/libsdl/app/SDLControllerManager.java#L347), and the way GUID was constructed in Godot's Linux joypad driver prior to the SDL3 joystick driver PR [here](https://github.com/godotengine/godot/blob/987832be460f82206fcac1467e58efbd111c5241/platform/linuxbsd/joypad_linux.cpp#L411).

<details>
<summary>DualShock 4 GUID before this PR</summary>

(Ignore that there's no fingerprint sensor here, it's a screenshot with a different PR applied)
![Screenshot_2025-12-25-18-17-15-579_org godotengine editor v4 debug](https://github.com/user-attachments/assets/6af143e6-5a40-4a63-b921-37ca221c2386)
</details>
<details>
<summary>DualShock 4 GUID after this PR + Input.get_joy_info()</summary>

![Screenshot_2025-12-25-21-29-55-945_org godotengine editor v4 debug](https://github.com/user-attachments/assets/52cdef91-9316-4524-a131-22d85aaac08d)
</details>

TODO:
- [x] Use SDL's internal mapping database
- [x] Add a note saying that changing the GUID-generating code to be different from SDL will make some GUIDs incompatible
- [ ] Fix swapped triggers issue by looking at SDL's Android code (for separate PR)
- [x] Fix triggers being mapped to [-1;1] on my Xbox One controller without breaking my Dualshock 4 triggers
- [x] If a mapping for an old GUID exists while a mapping for a new one doesn't, fallback to old mapping format
- [ ] Don't introduce regressions for Nvidia Shield axis mapping (for separate PR)
- [ ] Make the CI pass

Nintorch commented on platform/android/java/lib/src/main/java/org/godotengine/godot/input/GodotInputHandler.java:
```diff
@@ -395,7 +400,147 @@ public void onInputDeviceAdded(int deviceId) {
+		}
+	}
+
+	int getAxisMask(List<InputDevice.MotionRange> ranges) {
```
I should probably add a comment that says these 2 functions come from SDL source code

Nintorch replied:
`// getAxisMask() and getButtonMask() were taken from SDL source code to make joypad GUIDs compatible with SDL`

m4gr3d commented on platform/android/java/lib/src/main/java/org/godotengine/godot/input/GodotInputHandler.java:
```diff
@@ -395,7 +400,149 @@ public void onInputDeviceAdded(int deviceId) {
+		}
+	}
+
+	// getAxisMask() and getButtonMask() were borrowed from SDL source code to make joypad GUIDs compatible with SDL controller mappings.
```
What's the license for the SDL source code?

m4gr3d commented on platform/android/java/lib/src/main/java/org/godotengine/godot/input/GodotInputHandler.java:
```diff
@@ -395,7 +400,149 @@ public void onInputDeviceAdded(int deviceId) {
+	// getAxisMask() and getButtonMask() were borrowed from SDL source code to make joypad GUIDs compatible with SDL controller mappings.
+	int getAxisMask(List<InputDevice.MotionRange> ranges) {
+		// For compatibility, keep computing the axis mask like before,
+		// only really distinguishing 2, 4 and 6 axes.
```
Add a comment to describe what axes 2, 4, and 6 represent.

m4gr3d commented on platform/android/java/lib/src/main/java/org/godotengine/godot/input/GodotInputHandler.java:
```diff
@@ -395,7 +400,149 @@ public void onInputDeviceAdded(int deviceId) {
+			0xFFFFFFFF, // 14 -> ??
+			0xFFFFFFFF, // 15 -> ??
+			0xFFFFFFFF, // 16 -> ??
+		};
```
Those two can be constants and the method itself can be made static.

m4gr3d reviewed (commented):
Alongside manual testing, can you add unit tests to validate the added logic.

m4gr3d commented on platform/android/java/lib/src/main/java/org/godotengine/godot/input/GodotInputHandler.java:
```diff
@@ -369,6 +456,8 @@ public void onInputDeviceAdded(int deviceId) {
 		//Helps with creating new joypad mappings.
 		Log.i(TAG, "=== New Input Device: " + joystick.name);
 
+		ArrayList<InputDevice.MotionRange> axes_motion_ranges = new ArrayList<>();
```
```suggestion
		ArrayList<InputDevice.MotionRange> axesMotionRanges = new ArrayList<>();
```

m4gr3d reviewed (commented):
The code looks good to me.

@Alex2782 Can you help validating the implementation.

akien-mga commented on platform/android/android_input_handler.h:
```diff
@@ -103,4 +105,11 @@ class AndroidInputHandler {
+	// Generate GUID for compatibility with SDL controller mappings.
+	void generate_joy_guid(char *p_buf, int p_buf_len, int p_vendor_id, int p_product_id, int p_button_mask, int p_axis_mask) {
+		if (p_vendor_id != 0 || p_product_id != 0) {
+			snprintf(p_buf, p_buf_len, "05000000%04x0000%04x0000%04x%04x", BSWAP16(p_vendor_id), BSWAP16(p_product_id), BSWAP16(p_button_mask), BSWAP16(p_axis_mask));
```
If `050000005e040000130b0000ffff3f00` is present in both internal SDL and SDL_GameControllerDB, then it might be a bug that it's in SDL_GameControllerDB and it should likely be removed.

It seems like it was added in SDL_GameControllerDB in 2020 with https://github.com/mdqinc/SDL_GameControllerDB/commit/d1f0e1e798b6aac86c424fa999c985be9db72a6d, and later to SDL itself in 2021 with https://github.com/libsdl-org/SDL/commit/afd100f02b7689e3b1cf441f7abf40f0313f328b.

They're not too keen on receiving PRs from Godot contributors (understandably, as for several years we had our own non-SDL implementation and misunderstood the scope of that repo, and sent custom mappings that Godot needed but SDL itself didn't need). But I'll open an issue about it as I believe it's not a Godot specific issue, just one we happened to discover.

Alex2782 commented:
`This branch has conflicts that must be resolved`

I'll probably have time again on Monday when an Xbox controller needs to be tested on Android.


https://developer.android.com/reference/android/view/KeyEvent#KEYCODE_MEDIA_RECORD
  
I don't see `KEYCODE_MEDIA_RECORD` in the `JOYPAD_BUTTON_MASK_KEYS` array yet.  (for `misc1:b15`)
I think the share button isn't being recognized yet, or I'm missing something.

Alex2782 reviewed (commented):
All keys are recognized, but LT and RT are swapped.



```
Godot Engine v4.7.dev.gh-114338.b25248d22 (2026-02-06 11:08:17 UTC) - https://godotengine.org
OpenGL API OpenGL ES 3.2 V@0502.0 (GIT@193b2ee, I593c16c433, 1633593732) (Date:10/07/21) - Compatibility - Using Device: Qualcomm - Adreno (TM) 650

+ Found newly connected joypad #0: Xbox Series X Controller - 050000005e040000130b0000ffff3f00
{ "raw_name": "Xbox Wireless Controller 4416", "vendor_id": "1118", "product_id": "2835" }
```


I think this is coming from the header file, but when I tested your SDL-PR a few days ago using the Android SDL test app, everything was OK, so it seems it's not compatible with Godot.

` "050000005e040000130b0000ffff3f00,Xbox Series X Controller,a:b0,b:b1,back:b4,misc1:b15,dpdown:b12,dpleft:b13,dpright:b14,dpup:b11,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,",`



----------------------


And I think we need to change the order; `godotcontrollerdb.txt` needs to be at the very bottom, right? So that we can override it properly.

```python
controller_databases = [
    "gamecontrollerdb.txt",
    "godotcontrollerdb.txt",
    "../../thirdparty/sdl/joystick/SDL_gamepad_db.h",
]
```

Nintorch commented:
But that's still strange that the trigger mapping hasn't changed between the previous used mapping and the current one aside of the +, but now they're swapped 😅

Nintorch commented:
Ohh I see, thank you for testing! :D
I will see if I'm able to reproduce it with my Xbox controller too and I will see what I can do, maybe I will have to port more code from SDL.

Nintorch commented:
So I found some differences between how Godot handles Android joypad axes and how SDL handles them.
Godot checks for duplicate axes:
https://github.com/godotengine/godot/blob/40448082abe972daa179c64d7889cd88a8e7ea9c/platform/android/java/lib/src/main/java/org/godotengine/godot/input/GodotInputHandler.java#L389-L407

SDL doesn't:
https://github.com/libsdl-org/SDL/blob/4e2fd57e77fb4a28c0eeef0670fc4121cc2cf1f9/android-project/app/src/main/java/org/libsdl/app/SDLControllerManager.java#L230-L240
```java
                    List<InputDevice.MotionRange> ranges = joystickDevice.getMotionRanges();
                    Collections.sort(ranges, new RangeComparator());
                    for (InputDevice.MotionRange range : ranges) {
                        if ((range.getSource() & InputDevice.SOURCE_CLASS_JOYSTICK) != 0) {
                            if (range.getAxis() == MotionEvent.AXIS_HAT_X || range.getAxis() == MotionEvent.AXIS_HAT_Y) {
                                joystick.hats.add(range);
                            } else {
                                joystick.axes.add(range);
                            }
                        }
                    }
```

Currently in master my Xbox triggers are mapped to [-1;1]. If I remove the duplicate checking functionality, they work just fine (Dualshock 4 triggers too).
Maybe that's also the reason the triggers are swapped in Godot for you?

This functionality was added in https://github.com/godotengine/godot/pull/45771 to fix https://github.com/godotengine/godot/issues/45712 , but to me the original issue looks more like a mapping issue that needs to be checked in SDL testcontroller too, just in case. It also provides a way to setup a mapping for the device, maybe that would be helpful too!

I'm gonna add a commit to remove the duplicate checking functionality here. May I ask if you can test your controller with that commit too, please?

Or if it doesn't help, we can leave the duplicate checking functionality only for Nvidia Shield controller.

Nintorch committed:
Add Android SDL GUID support

This PR adds support for SDL-compatible joypad GUIDs on Android, so now we should be able to use SDL controller mapping database for Android controllers.
This PR adds support for Input.get_joy_info() for Android as well.

Nintorch commented:
I feel like the triggers issues should be fixed in a separate PR since this PR is about the mappings and GUID, but I feel like it should still be easy enough to fix.
For now I will need to make the CI pass.

Nintorch commented:
I think it's probably a better idea to integrate SDL joystick for Android later too, instead of merging a fix like this, since we'll need SDL joystick for Android for things like the HIDAPI driver, so I decided to close this PR.
